From 7d2e400418393a8153542441e88847db60f7ade7 Mon Sep 17 00:00:00 2001
From: Dan O'Donovan <danielx.o'donovan@intel.com>
Date: Thu, 29 May 2014 17:58:28 +0200
Subject: [PATCH 336/429] spi: Add platform data init handlers for spidev
 device registration

The Edison platform requires 1 instance of the spidev driver to be created
for SPI channel 5.1. This patch adds the platform data init handler required
to register this device correctly.

Signed-off-by: Dan O'Donovan <danielx.o'donovan@intel.com>
---
 arch/x86/platform/intel-mid/board.c           |  2 +
 .../platform/intel-mid/device_libs/Makefile   |  1 +
 .../intel-mid/device_libs/platform_spidev.c   | 40 +++++++++++++++++++
 .../intel-mid/device_libs/platform_spidev.h   | 20 ++++++++++
 4 files changed, 63 insertions(+)
 create mode 100644 arch/x86/platform/intel-mid/device_libs/platform_spidev.c
 create mode 100644 arch/x86/platform/intel-mid/device_libs/platform_spidev.h

diff --git a/arch/x86/platform/intel-mid/board.c b/arch/x86/platform/intel-mid/board.c
index 89a470000ee9..7fd6714a3536 100644
--- a/arch/x86/platform/intel-mid/board.c
+++ b/arch/x86/platform/intel-mid/board.c
@@ -79,6 +79,7 @@
  * SPI devices
  */
 #include "device_libs/platform_max3111.h"
+#include "device_libs/platform_spidev.h"
 
 /* WIFI devices */
 #include "device_libs/platform_wl12xx.h"
@@ -106,6 +107,7 @@ struct devs_id __initconst device_ids[] = {
 	{"pcal9555a-4", SFI_DEV_TYPE_I2C, 1, &pcal9555a_platform_data, NULL},
 
 	/* SPI devices */
+	{"spidev", SFI_DEV_TYPE_SPI, 0, &spidev_platform_data, NULL},
 	{"bma023", SFI_DEV_TYPE_I2C, 1, &no_platform_data, NULL},
 	{"pmic_gpio", SFI_DEV_TYPE_SPI, 1, &pmic_gpio_platform_data, NULL},
 	{"pmic_gpio", SFI_DEV_TYPE_IPC, 1, &pmic_gpio_platform_data,
diff --git a/arch/x86/platform/intel-mid/device_libs/Makefile b/arch/x86/platform/intel-mid/device_libs/Makefile
index a0a82f292642..4c0f02b8d9d7 100644
--- a/arch/x86/platform/intel-mid/device_libs/Makefile
+++ b/arch/x86/platform/intel-mid/device_libs/Makefile
@@ -38,6 +38,7 @@ obj-$(subst m,y,$(CONFIG_BQ24261_CHARGER)) += platform_bq24261.o
 obj-$(subst m,y,$(CONFIG_GPIO_PCA953X)) += platform_pcal9555a.o
 # SPI Devices
 obj-$(subst m,y,$(CONFIG_SERIAL_MRST_MAX3110)) += platform_max3111.o
+obj-$(subst m,y,$(CONFIG_SPI_SPIDEV)) += platform_spidev.o
 # MISC Devices
 obj-$(subst m,y,$(CONFIG_KEYBOARD_GPIO)) += platform_gpio_keys.o
 # ADC
diff --git a/arch/x86/platform/intel-mid/device_libs/platform_spidev.c b/arch/x86/platform/intel-mid/device_libs/platform_spidev.c
new file mode 100644
index 000000000000..a2b2ec3921af
--- /dev/null
+++ b/arch/x86/platform/intel-mid/device_libs/platform_spidev.c
@@ -0,0 +1,40 @@
+/*
+ * platform_spidev.c: spidev platform data initilization file
+ *
+ * (C) Copyright 2014 Intel Corporation
+ * Author: Dan O'Donovan <dan@emutex.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2
+ * of the License.
+ */
+
+#include <linux/spi/spi.h>
+#include <linux/spi/intel_mid_ssp_spi.h>
+#include <asm/intel-mid.h>
+#include "platform_spidev.h"
+
+static struct intel_mid_ssp_spi_chip chip = {
+	.burst_size = DFLT_FIFO_BURST_SIZE,
+	.timeout = DFLT_TIMEOUT_VAL,
+	/* SPI DMA is currently not usable on Tangier */
+	.dma_enabled = false,
+};
+
+void __init *spidev_platform_data(void *info)
+{
+	struct spi_board_info *spi_info = info;
+
+	if (!spi_info) {
+		pr_err("%s: invalid info pointer\n", __func__);
+		return NULL;
+	}
+
+	spi_info->mode = SPI_MODE_0;
+
+	spi_info->controller_data = &chip;
+	spi_info->bus_num = FORCE_SPI_BUS_NUM;
+
+	return NULL;
+}
diff --git a/arch/x86/platform/intel-mid/device_libs/platform_spidev.h b/arch/x86/platform/intel-mid/device_libs/platform_spidev.h
new file mode 100644
index 000000000000..a40ef8ff29a6
--- /dev/null
+++ b/arch/x86/platform/intel-mid/device_libs/platform_spidev.h
@@ -0,0 +1,20 @@
+/*
+ * platform_spidev.h: spidev platform data header file
+ *
+ * (C) Copyright 2014 Intel Corporation
+ * Author: Dan O'Donovan <dan@emutex.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2
+ * of the License.
+ */
+#ifndef _PLATFORM_SPIDEV_H_
+#define _PLATFORM_SPIDEV_H_
+
+/* REVERT ME workaround[MRFL] for invalid bus number in IAFW .25 */
+#define FORCE_SPI_BUS_NUM	5
+#define FORCE_CHIP_SELECT	1
+
+extern void *spidev_platform_data(void *info) __attribute__((weak));
+#endif
-- 
2.37.3

